rmf_task_ros2/test/action/test_ActionInterface.cpp

```diff
@@ -0,0 +1,194 @@
+
+namespace rmf_task_ros2 {
+namespace action {
+
+//==============================================================================
+SCENARIO("Action communication with client and server", "[ActionInterface]")
+{
+  TaskProfile task_profile1;
+  task_profile1.task_id = "task1";
+  task_profile1.task_type.type = TaskType::TYPE_STATION;
+
+  TaskProfile task_profile2;
+  task_profile2.task_id = "task2";
+  task_profile2.task_type.type = TaskType::TYPE_STATION;
+
+//==============================================================================
+
+  // received task to test
+  rmf_utils::optional<TaskProfile> test_add_task;
+  rmf_utils::optional<TaskProfile> test_cancel_task;
+
+  // Creating 1 auctioneer and 1 bidder
+  rclcpp::init(0, nullptr);
+  auto node = rclcpp::Node::make_shared("test_ActionInferface");
+
+  using DispatchRequest = rmf_task_msgs::msg::DispatchRequest;
+  using TaskSummary = rmf_task_msgs::msg::TaskSummary;
+
+  auto action_server = TaskActionServer<DispatchRequest, TaskSummary>::make(
+    node, "test_server");
+  auto action_client = TaskActionClient<DispatchRequest, TaskSummary>::make(
+    node);
+
+  rclcpp::executors::SingleThreadedExecutor executor;
+  executor.add_node(node);
+
+  // received test request msg from client
+  action_server->register_callbacks(
+    // add task
+    [&test_add_task](const TaskProfile& task_profile)
+    {
+      test_add_task = task_profile;
+      return true;
+    },
+    // cancel task
+    [&test_cancel_task](const TaskProfile& task_profile)
+    {
+      test_cancel_task = task_profile;
+      return true;
+    }
+  );
+
+  // ROS Spin: forever incompleted future
+  std::promise<void> ready_promise;
+  std::shared_future<void> ready_future(ready_promise.get_future());
+
+  WHEN("Add Task")
+  {
+    // Add invalid Task!
+    TaskStatusPtr status_ptr(new TaskStatus);
+
+    action_client->add_task("wrong_server", task_profile1, status_ptr);
+
+    executor.spin_until_future_complete(ready_future,
+      rmf_traffic::time::from_seconds(0.5));
+
+    // should not receive cuz incorrect serverid
+    REQUIRE(status_ptr->state == TaskStatus::State::Pending);
+
+    action_client->add_task("test_server", task_profile1, status_ptr);
+    executor.spin_until_future_complete(ready_future,
+      rmf_traffic::time::from_seconds(0.5));
+
+    // check status
+    REQUIRE(status_ptr->state == TaskStatus::State::Queued);
+
+    // status ptr is destroyed, should not have anymore tracking
+    status_ptr.reset();
+    REQUIRE(action_client->size() == 0);
+  }
+
+  WHEN("Cancel Task")
+  {
+    // send valid task
+    TaskStatusPtr status_ptr(new TaskStatus);
+    action_client->add_task("test_server", task_profile2, status_ptr);
+
+    executor.spin_until_future_complete(ready_future,
+      rmf_traffic::time::from_seconds(0.5));
+
+    // Invalid Cancel Task!
+    bool cancel_success = action_client->cancel_task(task_profile1);
+    executor.spin_until_future_complete(ready_future,
```

Comment:
I think you could just use [`spin_some`](https://github.com/ros2/rclcpp/blob/543a3c39c16fdeb3f0109d0e906105870f547fe1/rclcpp/include/rclcpp/executor.hpp#L296) here.

Reply:
The reason for using `spin_until_future_complete` here is to spin with a timeout duration arg (without looping it).

Reply:
Oh right, the duration argument for `spin_some` is a timeout, and it might quit earlier. Okay, we can keep this as-is, although I'm tempted to request a more direct API for this from `rclcpp`.